Cabal/src/Distribution/PackageDescription/Check.hs

```diff
@@ -2412,27 +2434,141 @@ checkGlobFiles :: Verbosity
+                                            extraDocFilesSupport
+                                            field wrongPaths)
+                                       | not (null wrongPaths) ]
+              in
+                ( if noMatchesWarn
+                    then [PackageDistSuspiciousWarn (GlobNoMatch field glob)] ++
+                         individualWarn ++
+                         wrongFieldWarnings
+                    else individualWarn ++ wrongFieldWarnings
+                , docFiles1'
+                )
+          where
+            checkGlobResult :: GlobResult FilePath
+                            -> ([PackageCheck], Bool, [FilePath], [FilePath])
+                            -> ([PackageCheck], Bool, [FilePath], [FilePath])
+            checkGlobResult result (ws, noMatchesWarn, docFiles2, wrongPaths) =
+              let noMatchesWarn' = noMatchesWarn &&
+                                   not (suppressesNoMatchesWarning result)
+              in case getWarning field glob result of
+                -- No match: add warning and do no further check
+                Left w ->
+                  ( w : ws
+                  , noMatchesWarn'
+                  , docFiles2
+                  , wrongPaths
+                  )
+                -- Match: check doc files
+                Right path ->
+                  let path' = makeRelative root (normalise path)
+                      (docFiles2', wrongPaths') = checkDoc isDocField
+                                                           path'
+                                                           docFiles2
+                                                           wrongPaths
+                  in
+                    ( ws
+                    , noMatchesWarn'
+                    , docFiles2'
+                    , wrongPaths'
+                    )
+
+    -- Check whether a path is a desirable doc: if so, check if it is in the
+    -- field "extra-doc-files".
+    checkDoc :: Bool                     -- Is it "extra-doc-files" ?
+             -> FilePath                 -- Path to test
+             -> [FilePath]               -- Pending doc files to check
+             -> [FilePath]               -- Previous wrong paths
+             -> ([FilePath], [FilePath]) -- Updated paths
+    checkDoc isDocField path docFiles wrongFieldPaths =
+      if path `elem` docFiles
+        -- Found desirable doc file
+        then
+          ( delete path docFiles
+          , if isDocField then wrongFieldPaths else path : wrongFieldPaths
+          )
+        -- Not a desirable doc file
+        else
+          ( docFiles
+          , wrongFieldPaths
+          )
+
+    -- Predicate for desirable documentation file on Hackage server
+    isDesirableExtraDocFile :: [FilePath] -> FilePath -> Bool
+    isDesirableExtraDocFile paths path = map toLower basename `elem` paths
+      where
+        (basename, _ext) = splitExtension path
```

Comment:
Extensions like `.bak`, `tmp`, `.*~`, files ending in `~` etc. should be ruled out here.

Reply:
.swp, .swn, swo,...

Almost better to whitelist extensions (no extension, `txt`, `md`, `rst`).

Reply:
also .markdown

Reply:
See: https://github.com/haskell/hackage-server/blob/master/src/Distribution/Server/Packages/ChangeLog.hs

I would also add rst. PR comming.

Reply:
See #8747